packages/expo-permissions/ios/EXPermissions/EXUserNotificationRequester.m

```diff
@@ -0,0 +1,103 @@
+@property (nonatomic, weak) id<EXPermissionRequesterDelegate> delegate;
+@property (nonatomic, weak) EXModuleRegistry * moduleRegistry;
+
+@end
+
+@implementation EXUserNotificationRequester
+
+- (instancetype)initWithModuleRegistry: (EXModuleRegistry *) moduleRegistry {
+  if (self = [super init]) {
+    _moduleRegistry = moduleRegistry;
+  }
+  return self;
+}
+
++ (id<EXUserNotificationCenterProxyInterface>)getCenterWithModuleRegistry:(EXModuleRegistry *) moduleRegistry {
+  return [moduleRegistry getModuleImplementingProtocol:@protocol(EXUserNotificationCenterProxyInterface)];
+}
+
++ (NSDictionary *)permissionsWithModuleRegistry:(EXModuleRegistry *)moduleRegistry
+{
+  dispatch_semaphore_t sem = dispatch_semaphore_create(0);
+  __block BOOL allowsSound;
+  __block BOOL allowsAlert;
+  __block BOOL allowsBadge;
+  __block EXPermissionStatus status;
+
+  [[EXUserNotificationRequester getCenterWithModuleRegistry:moduleRegistry] getNotificationSettingsWithCompletionHandler:^(UNNotificationSettings *settings) {
+    allowsSound = settings.soundSetting == UNNotificationSettingEnabled;
+    allowsAlert = settings.alertSetting == UNNotificationSettingEnabled;
+    allowsBadge = settings.badgeSetting == UNNotificationSettingEnabled;
+
+    status = EXPermissionStatusUndetermined;
+
+    if (settings.authorizationStatus == UNAuthorizationStatusAuthorized) {
+      status = EXPermissionStatusGranted;
+    } else if (settings.authorizationStatus == UNAuthorizationStatusDenied) {
+      status = EXPermissionStatusDenied;
+    } else {
+      status = EXPermissionStatusUndetermined;
+    }
+    dispatch_semaphore_signal(sem);
+  }];
+
+  dispatch_semaphore_wait(sem, DISPATCH_TIME_FOREVER);
+
+  return @{
+           @"status": [EXPermissions permissionStringForStatus:status],
+           @"allowsSound": @(allowsSound),
+           @"allowsAlert": @(allowsAlert),
+           @"allowsBadge": @(allowsBadge),
+           @"expires": EXPermissionExpiresNever,
+           };
+}
+
+- (void)setDelegate:(id<EXPermissionRequesterDelegate>)delegate
+{
+  _delegate = delegate;
+}
+
+- (void)requestPermissionsWithResolver:(EXPromiseResolveBlock)resolve rejecter:(EXPromiseRejectBlock)reject
+{
+  if (_resolve != nil || _reject != nil) {
+    reject(@"E_AWAIT_PROMISE", @"Another request for the same permission is already being handled.", nil);
+    return;
+  }
+
+  _resolve = resolve;
+  _reject = reject;
+
+  __weak EXUserNotificationRequester *weakSelf = self;
+
+  UNAuthorizationOptions options = UNAuthorizationOptionAlert + UNAuthorizationOptionSound + UNAuthorizationOptionBadge;
+  id<EXUserNotificationCenterProxyInterface> notificationCenter = [EXUserNotificationRequester getCenterWithModuleRegistry:_moduleRegistry];
+  [notificationCenter requestAuthorizationWithOptions:options completionHandler:^(BOOL granted, NSError * _Nullable error) {
+    dispatch_async(dispatch_get_main_queue(), ^{
+      [weakSelf _consumeResolverWithCurrentPermissions];
```

Comment:
Instead of storing `resolve` and `reject` as instance properties, I think it would be better to have a separate property for whether we are currently requesting permissions. That way, `resolve` and `reject` can't be called from other places in this class:

 ```objc
strongSelf = weakSelf
if (!strongSelf) {
  return;
}

strongSelf->_isRequestingPermissions = NO;

if (error) {  // should handle errors too
  reject(...);
  return;
}

resolve([strongSelf class] permissionsWithModuleRegistry:_moduleRegistry])
if (strongSelf->_delegate) {
  ...
}
```

Reply:
The difference is that `resolve` and `reject` are kept within the scope of the code that provided them (the caller of `-[requestPermissionsWithResolver:(EXPromiseResolveBlock)resolve rejecter:(EXPromiseRejectBlock)reject]`) and  one call to `requestPermissionsWithResolver` cannot access the `resolve` and `reject` blocks of another call to `requestPermissionsWithResolver`.

I think it'd also achieve the same thing to add `resolve` as a parameter of `_consumeResolverWithCurrentPermissions`. 

If this class becomes more complex, it will be harder to ensure that we call the correct `resolve` and `reject` blocks if they are stored as instance properties. Sometimes using the instance properties is necessary but in this case I think having fewer properties would be better. The point is to not expose `resolve` and `reject` as instance properties -- the state we need to store in an instance property is only the fact that a concurrent call to `requestPermissionsWithResolver` is happening, and we don't need the power to resolve/reject that concurrent call to be exposed to other calls.